Sources/Models/RequestModel.swift

```diff
@@ -11,24 +11,77 @@ import UIKit
 open class RequestModel: Codable {
     public let id: String
     public let url: String
+    public let host: String?
+    public let port: Int?
+    public let scheme: String?
     public let date: Date
     public let method: String
-    public let headers: [String: String]?
+    public let headers: [String: String]
+    public var credentials: [String : String]
+    public var cookies: String?
     open var httpBody: Data?
     open var code: Int
     open var responseHeaders: [String: String]?
     open var dataResponse: Data?
     open var errorClientDescription: String?
     open var duration: Double?
     
-    init(request: NSURLRequest) {
+    init(request: NSURLRequest, session: URLSession?) {
         id = UUID().uuidString
         url = request.url?.absoluteString ?? ""
+        host = request.url?.host
+        port = request.url?.port
+        scheme = request.url?.scheme
         date = Date()
-        method = request.httpMethod ?? "N/A"
-        headers = request.allHTTPHeaderFields
+        method = request.httpMethod ?? "GET"
+        credentials = [:]
+        var headers = request.allHTTPHeaderFields ?? [:]
         httpBody = request.httpBody
         code = 0
+        
+        
+        // collect all HTTP Request headers except the "Cookie" header. Many request representations treat cookies with special parameters or structures. For cookie collection, refer to the bottom part of this method
+        session?.configuration.httpAdditionalHeaders?
+            .filter {  $0.0 != AnyHashable("Cookie") }
+            .forEach { element in
+                guard let key = element.0 as? String, let value = element.1 as? String else { return }
+                headers[key] = value
+        }
+        self.headers = headers
+        
+        // if the target server uses HTTP Basic Authentication, collect username and password
+        if let credentialStorage = session?.configuration.urlCredentialStorage,
+            let host = self.host,
+            let port = self.port {
+            let protectionSpace = URLProtectionSpace(
+                host: host,
+                port: port,
+                protocol: scheme,
+                realm: host,
+                authenticationMethod: NSURLAuthenticationMethodHTTPBasic
+            )
+
+            if let credentials = credentialStorage.credentials(for: protectionSpace)?.values {
+                for credential in credentials {
+                    guard let user = credential.user, let password = credential.password else { continue }
+                    self.credentials[user] = password
+                }
+            }
+        }
+        
+        //  collect cookies associated with the target host
+        //  TODO: Add the else branch.
```

Comment:
@smarttuner do you remember what this todo refers to? What should the else do?

Reply:
This todo is referred to the if condition placed at line 79. The else branch needs to pick Cookies from other sources. Reading [https://developer.apple.com/documentation/foundation/urlsessionconfiguration/1411589-httpshouldsetcookies](https://developer.apple.com/documentation/foundation/urlsessionconfiguration/1411589-httpshouldsetcookies) , this information can be part of session’s httpAdditionalHeaders (so we can assign to self.cookies the value of httpAdditionalHeaders["Cookie"]) or can be part of a custom implementation of NSURLRequest.
If we want to handle only the first case, we can rework the if condition as follow:
`if let session = session {
            if session.configuration.httpShouldSetCookies {
                if let url = request.url, let cookieStorage = session.configuration.httpCookieStorage,
                    let cookies = cookieStorage.cookies(for: url), !cookies.isEmpty {
                    self.cookies = cookies.reduce("") { $0 + "\($1.name)=\($1.value);" }
                }
            }
            else {
                self.cookies = session.configuration.httpAdditionalHeaders?["Cookie"] as? String
            }
        }
`